webrender/src/device/gfx/device.rs

```diff
@@ -1882,6 +1911,114 @@ impl<B: hal::Backend> Device<B> {
         rbo_id
     }
 
+    fn create_cache_buffer(
+        &mut self,
+        width: u64,
+        height: u64,
+    ) -> PMBuffer<B> {
+        let buffer_stride = std::mem::size_of::<[f32; 4]>() as u64;
+        let buffer_len = height * width * buffer_stride;
+
+        assert_ne!(buffer_len, 0);
+        use hal::buffer::Usage as BU;
+        let mut storage_buffer =
+            unsafe { self.device.create_buffer(buffer_len, BU::STORAGE | BU::TRANSFER_SRC | BU::TRANSFER_DST) }.unwrap();
+
+        let buffer_req = unsafe { self.device.get_buffer_requirements(&storage_buffer) };
+        let mut coherent = false;
+
+        use hal::memory::Properties;
+        let memory_types = self.adapter.physical_device.memory_properties().memory_types;
+        let upload_type = memory_types
+            .iter()
+            .enumerate()
+            .position(|(id, mem_type)| {
+                // type_mask is a bit field where each bit represents a memory type. If the bit is set
+                // to 1 it means we can use that type for our buffer. So this code finds the first
+                // memory type that has a `1` (or, is allowed), and is visible to the CPU.
+                buffer_req.type_mask & (1 << id) != 0
+                    && mem_type.properties.contains(Properties::CPU_VISIBLE | Properties::COHERENT)
```

Comment:
I think the implementation guarantees that such a memory type exists, i.e. for any usage of a buffer, there has to be a coherent memory type that supports it

Reply:
That's right `There must be at least one memory type with both the VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT and VK_MEMORY_PROPERTY_HOST_COHERENT_BIT bits set in its propertyFlags`

Reply:
Fixed it in https://github.com/szeged/webrender/pull/297/commits/76b255b4b7ec31e659242e04dbd757177777abf4#diff-cdf03c96bd34cb98c2d09b0be7658232R1980

Reply:
Fixed it in https://github.com/szeged/webrender/pull/297/commits/76b255b4b7ec31e659242e04dbd757177777abf4#diff-cdf03c96bd34cb98c2d09b0be7658232R1980